matrix/rules.py

```diff
@@ -368,8 +369,13 @@ def allow_event(e):
                 self.handle_shutdown(None)
                 if self.fail_fast and e.task.gating is True:
                     log.info("Fail Fast on Test failure: %s" % test.name)
+                    self.exit_code = 1
```

Comment:
This check isn't actually catching our TestFailure, so setting the .exit_code here doesn't work. :-/

Need to investigate and fix ...